File: swift_reference/Views/JoinCircleView.swift
import SwiftUI

struct JoinCircleView: View {
    @Environment(\.dismiss) private var dismiss

    @State private var inviteCode: String
    @State private var isLoading: Bool = false
    @State private var error: String?
    @State private var joinedName: String?

    var onJoined: () async -> Void

    init(initialCode: String = "", onJoined: @escaping () async -> Void) {
        _inviteCode = State(initialValue: initialCode)
        self.onJoined = onJoined
    }

    var body: some View {
        NavigationStack {
            VStack(spacing: 24) {
                if let joinedName {
                    VStack(spacing: 16) {
                        Spacer()

                        ZStack {
                            Circle()
                                .fill(TributeColor.sage.opacity(0.12))
                                .frame(width: 80, height: 80)
                            Image(systemName: "checkmark.circle.fill")
                                .font(.system(size: 36))
                                .foregroundStyle(TributeColor.sage)
                        }

                        Text("Joined \(joinedName)!")
                            .font(.system(.title3, design: .serif, weight: .semibold))
                            .foregroundStyle(TributeColor.warmWhite)

                        Button {
                            Task {
                                await onJoined()
                                dismiss()
                            }
                        } label: {
                            Text("Done")
                                .tributeButton()
                        }
                        .padding(.horizontal, 20)

                        Spacer()
                    }
                } else {
                    VStack(alignment: .leading, spacing: 16) {
                        VStack(alignment: .leading, spacing: 6) {
                            Text("Invite Code")
                                .font(.caption.weight(.semibold))
                                .foregroundStyle(.secondary)
                            TextField("Enter invite code", text: $inviteCode)
                                .textFieldStyle(.plain)
                                .textInputAutocapitalization(.characters)
                                .autocorrectionDisabled()
                                .padding(12)
                                .background(TributeColor.cardBackground)
                                .clipShape(.rect(cornerRadius: 10))
                                .overlay(
                                    RoundedRectangle(cornerRadius: 10)
                                        .strokeBorder(TributeColor.cardBorder, lineWidth: 0.5)
                                )
                        }

                        Text("Ask the circle creator for their invite code, or tap a shared invite link.")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }

                    if let error {
                        HStack(spacing: 6) {
                            Image(systemName: "exclamationmark.triangle.fill")
                                .font(.caption)
                            Text(error)
                                .font(.caption)
                        }
                        .foregroundStyle(TributeColor.warmCoral)
                    }

                    Button {
                        Task { await joinCircle() }
                    } label: {
                        Group {
                            if isLoading {
                                ProgressView()
                                    .tint(TributeColor.charcoal)
                            } else {
                                Text("Join Circle")
                            }
                        }
                        .tributeButton()
                    }
                    .disabled(inviteCode.trimmingCharacters(in: .whitespaces).isEmpty || isLoading)
                    .opacity(inviteCode.trimmingCharacters(in: .whitespaces).isEmpty ? 0.5 : 1)

                    Spacer()
                }
            }
            .padding(20)
            .background {
                ZStack {
                    TributeColor.charcoal.ignoresSafeArea()
                    TributeColor.warmGlow.ignoresSafeArea()
                }
            }
            .navigationTitle("Join Circle")
            .navigationBarTitleDisplayMode(.inline)
            .toolbarColorScheme(.dark, for: .navigationBar)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                        .foregroundStyle(.secondary)
                }
            }
        }
    }

    private func joinCircle() async {
        let code = inviteCode.trimmingCharacters(in: .whitespaces)
        guard !code.isEmpty else { return }

        isLoading = true
        error = nil

        do {
            let response = try await APIService.shared.joinCircle(inviteCode: code)
            if response.alreadyMember {
                error = "You're already a member of this circle"
            } else {
                joinedName = response.name
            }
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }
}

File: lib/views/circles/join_circle_view.dart
import 'package:flutter/material.dart';
import '../../services/api_service.dart';
import '../../theme/app_theme.dart';

class JoinCircleView extends StatefulWidget {
  final String? initialCode;
  final Future<void> Function()? onJoined;

  const JoinCircleView({super.key, this.initialCode, this.onJoined});

  @override
  State<JoinCircleView> createState() => _JoinCircleViewState();
}

class _JoinCircleViewState extends State<JoinCircleView> {
  late final TextEditingController _codeController;
  bool _isLoading = false;
  String? _error;
  String? _joinedName;

  @override
  void initState() {
    super.initState();
    _codeController = TextEditingController(text: widget.initialCode ?? '');
  }

  @override
  void dispose() {
    _codeController.dispose();
    super.dispose();
  }

  Future<void> _joinCircle() async {
    final code = _codeController.text.trim().toUpperCase();
    if (code.isEmpty) return;

    setState(() { _isLoading = true; _error = null; });
    try {
      final response = await APIService.shared.joinCircle(code);
      if (!mounted) return;
      if (response.alreadyMember) {
        setState(() { _error = "You're already a member of this circle"; _isLoading = false; });
      } else {
        setState(() { _joinedName = response.name; _isLoading = false; });
      }
    } catch (e) {
      if (mounted) setState(() { _error = e.toString(); _isLoading = false; });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: TributeColor.charcoal,
      appBar: AppBar(
        backgroundColor: TributeColor.charcoal,
        title: const Text('Join Circle',
            style: TextStyle(color: TributeColor.warmWhite, fontSize: 17, fontWeight: FontWeight.w600)),
        leading: TextButton(
          onPressed: () => Navigator.pop(context),
          child: Text('Cancel', style: TextStyle(color: Colors.white.withValues(alpha: 0.5))),
        ),
        leadingWidth: 80,
      ),
      body: SafeArea(
        child: Padding(
          padding: const EdgeInsets.all(20),
          child: _joinedName != null ? _successState() : _formState(),
        ),
      ),
    );
  }

  Widget _successState() {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Container(
          width: 80, height: 80,
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            color: TributeColor.sage.withValues(alpha: 0.12),
          ),
          child: const Icon(Icons.check_circle_rounded, size: 36, color: TributeColor.sage),
        ),
        const SizedBox(height: 16),
        Text('Joined $_joinedName!',
            style: const TextStyle(fontSize: 20, fontWeight: FontWeight.w600, color: TributeColor.warmWhite)),
        const SizedBox(height: 24),
        SizedBox(
          width: double.infinity,
          child: ElevatedButton(
            onPressed: () async {
              await widget.onJoined?.call();
              if (mounted) Navigator.pop(context);
            },
            style: ElevatedButton.styleFrom(
              backgroundColor: TributeColor.golden,
              foregroundColor: TributeColor.charcoal,
              padding: const EdgeInsets.symmetric(vertical: 16),
              shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(14)),
            ),
            child: const Text('Done', style: TextStyle(fontWeight: FontWeight.w600, fontSize: 16)),
          ),
        ),
      ],
    );
  }

  Widget _formState() {
    final isEmpty = _codeController.text.trim().isEmpty;
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        const Text('Invite Code',
            style: TextStyle(fontSize: 12, fontWeight: FontWeight.w600, color: TributeColor.softGold)),
        const SizedBox(height: 6),
        TextField(
          controller: _codeController,
          onChanged: (_) => setState(() {}),
          textCapitalization: TextCapitalization.characters,
          autocorrect: false,
          style: const TextStyle(color: TributeColor.warmWhite),
          decoration: InputDecoration(
            hintText: 'Enter invite code',
            hintStyle: TextStyle(color: Colors.white.withValues(alpha: 0.3)),
            filled: true,
            fillColor: TributeColor.cardBackground,
            border: OutlineInputBorder(
              borderRadius: BorderRadius.circular(10),
              borderSide: BorderSide(color: TributeColor.cardBorder, width: 0.5),
            ),
            enabledBorder: OutlineInputBorder(
              borderRadius: BorderRadius.circular(10),
              borderSide: BorderSide(color: TributeColor.cardBorder, width: 0.5),
            ),
            focusedBorder: OutlineInputBorder(
              borderRadius: BorderRadius.circular(10),
              borderSide: BorderSide(color: TributeColor.golden.withValues(alpha: 0.5), width: 1),
            ),
            contentPadding: const EdgeInsets.all(12),
          ),
        ),
        const SizedBox(height: 8),
        Text(
          'Ask the circle creator for their invite code, or tap a shared invite link.',
          style: TextStyle(fontSize: 12, color: Colors.white.withValues(alpha: 0.4)),
        ),
        if (_error != null) ...[
          const SizedBox(height: 12),
          Row(children: [
            const Icon(Icons.warning_amber, size: 14, color: TributeColor.warmCoral),
            const SizedBox(width: 6),
            Expanded(child: Text(_error!, style: const TextStyle(fontSize: 12, color: TributeColor.warmCoral))),
          ]),
        ],
        const SizedBox(height: 24),
        Opacity(
          opacity: isEmpty ? 0.5 : 1.0,
          child: SizedBox(
            width: double.infinity,
            child: ElevatedButton(
              onPressed: (isEmpty || _isLoading) ? null : _joinCircle,
              style: ElevatedButton.styleFrom(
                backgroundColor: TributeColor.golden,
                foregroundColor: TributeColor.charcoal,
                padding: const EdgeInsets.symmetric(vertical: 16),
                shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(14)),
              ),
              child: _isLoading
                  ? const SizedBox(width: 18, height: 18,
                      child: CircularProgressIndicator(strokeWidth: 2, color: TributeColor.charcoal))
                  : const Text('Join Circle', style: TextStyle(fontWeight: FontWeight.w600, fontSize: 16)),
            ),
          ),
        ),
      ],
    );
  }
}
